### PR TITLE
Fix doc comments on hover for methods with params starting with 'end'

### LINKED ISSUE #6676: Doc comments not showing up on hover for certain methods
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0A%23%20docs%20that%20should%20show%20up%20on%20hover%0A%23%20...%20they%20show%20up%20if%20you%20delete%20%60end_time%60%20as%20a%20parameter%0A%23%20Are%20we%20doing%20some%20source-parsing%20heuristics%20that%20get%20in%20the%20way%3F%0Asig%20do%0A%20%20params%28%0A%20%20%20%20merchant%3A%20String%2C%0A%20%20%20%20product_features%3A%20T%3A%3AArray%5BString%5D%2C%0A%20%20%20%20start_time%3A%20Integer%2C%0A%20%20%20%20end_time%3A%20T.nilable%28Integer%29%2C%0A%20%20%29%0A%20%20%20%20.void%0Aend%0Adef%20example%28%0A%20%20merchant%3A%2C%0A%20%20product_features%3A%2C%0A%20%20start_time%3A%2C%0A%20%20end_time%3A%0A%29%0A%20%20puts%28merchant%29%0Aend%0A)

```ruby
# typed: true
extend T::Sig

# docs that should show up on hover
# ... they show up if you delete `end_time` as a parameter
# Are we doing some source-parsing heuristics that get in the way?
sig do
  params(
    merchant: String,
    product_features: T::Array[String],
    start_time: Integer,
    end_time: T.nilable(Integer),
  )
    .void
end
def example(
  merchant:,
  product_features:,
  start_time:,
  end_time:
)
  puts(merchant)
end

```

#### Observed output

```
No errors! Great job.
```

When you hover on the `call` in `def call`, the doc comment doesn't appear in the hover text.

#### Expected behavior

The doc comment should appear in the hover text.

- - -

Note that if you delete the `end_time` variable from the sig and the method def, the doc comment shows up in the hover text. I wonder if we're doing something artificially limiting like counting lines limiting when trying to find a method's comment?

### PR DESCRIPTION
**Pull request by komarovb**
This fixes a specific issue that occurs only when a valid line starts with an 'end' within the sig do/end block.

### Motivation

Fixes #6676 

### Test plan

See included automated tests.

### CODE REVIEW

File: main/lsp/lsp_helpers.cc [hunk 1]
```diff
@@ -332,7 +332,7 @@ optional<string> findDocumentation(string_view sourceCode, int beginIndex) {
                 // Start of sig block
                 && !(absl::StartsWith(line, "sig do") || absl::StartsWith(line, "sig(:final) do"))
                 // Invalid end keyword
-                && !absl::StartsWith(line, "end")) {
+                && !(absl::StartsWith(line, "end") && absl::EndsWith(line, "end"))) {
```
Comment (jez) [inline main/lsp/lsp_helpers.cc:335]: Since `line` has been initialized with `StripAsciiWhitespace`, we can just check `line != "end"`.

Doing it the way this PR currently implements it with "StartsWith + EndsWith" means that the bug in #6676 could still happen for a signature like

```ruby
sig do
  params(
    end_time: Time_end
  ).void
end
def foo(end_time)
end
```

**Review by jez — APPROVED** [review]

**Review by jez — CHANGES_REQUESTED** [review]
^ sorry, wrong button

**Comment by jez** [comment]
Approach mostly looks good but let's make sure we handle the edge cases.

File: main/lsp/lsp_helpers.cc (hunk 1 above)
Comment (komarovb) [reply main/lsp/lsp_helpers.cc:335]: Good point. My mistake, thank you for reviewing this.

Comment (jez) [reply main/lsp/lsp_helpers.cc:335]: no worries, this code is cursed anyways. Thanks for the contribution!

Comment (komarovb) [reply main/lsp/lsp_helpers.cc:335]: Just out of curiosity, is there any particular reason why you need to handle this "invalid end keyword" case here? 

I can think of somebody breaking current logic just by putting something like an `end # random` line somewhere in the `sig` block.

Comment (jez) [reply main/lsp/lsp_helpers.cc:335]: The fundamental problem here is that Sorbet doesn't parse comments in its AST.

https://github.com/sorbet/sorbet/blob/ff9bec08e7c2f5404f326534d52b94002e39765e/parser/parser/cc/lexer.rl#L275-L281

There is a callback that could be implemented if Sorbet wanted to start doing that, but due to memory constraints running Sorbet over Stripe's main Ruby codebase, we've been reluctant to do anything which would cause Sorbet to use more memory.

If Sorbet were able to access the list of already-parsed comments in its main pipeline, we would not need to resort to ad hoc string scanning inside LSP helpers like this.

**tl;dr**: The current solution is a compromise so that Sorbet can use substantially less memory while sometimes failing to find a documentation comment and sometimes finding the wrong comment. Given that this information is not load boarding, a best-effort solution is mostly fine for now.

Comment (komarovb) [reply main/lsp/lsp_helpers.cc]: Thank you for such a detailed reply! 

I pushed the requested changes.

**Review by jez — APPROVED** [review]
